### PR TITLE
Add ability to observe expect/send errors.  Useful for testing.

### PR DESCRIPTION
**Pull request by coryb**
We need a way to observe the success/failures of Expect/Send operations for testing.  I think the observer pattern will work well for this.  I added two more With ConsoleOpts to allow users to register callbacks on Expect results and Send results.   I have also modified all the tests to assert that there are no failures from `c.Expect` or `c.Send` calls. (I also added checks to verify Close calls are not silently failing also).

I did find some expect statements in the tests were failing due to race conditions of goroutines, so I have updated them to use `sync.WaitGroup` to ensure everything is run before we exit the unit test.

I did modify the Matcher interface to add a `Criteria() interface{}` function.  This is useful for the ExpectObserver callbacks so they can inspect that the matcher was trying to match, example usage:
```go
if err != nil {
    t.Errorf("Failed to find %v in %q: %s", matcher.Criteria(), buf, err)
}
```

### CODE REVIEW

File: expect_test.go [hunk 1]
```diff
@@ -84,18 +125,24 @@ func TestExpect(t *testing.T) {
 	if err != nil {
 		t.Errorf("Expected no error but got '%s'", err)
 	}
+	// close the pts so we can expect EOF
+	testCloser(t, c.Tty())
```
Comment (coryb) [inline expect_test.go]: If we call `c.Close()` first then there is a race condition where the goroutine will be trying to flush the response via `runeWriter.Flush()` or waiting for an EOF, but the ptm is closed causing `os.PathError`.  Closing the `c.Tty()` will correctly trigger the EOF for the goroutine, which will allow it to exit properly without errors.

**Comment by coryb** [comment]
I ran golint on it:
```
console.go:59:6:warning: type name will be used as expect.ExpectObserver by other packages, and that stutters; consider calling this Observer (golint)
```

Not sure if you have other suggestions for naming things.

**Comment by coryb** [comment]
Looking into test failures.  It seems when closing the pts on OSX it works fine, the ExpectEOF calls actually gets an EOF.  On Linux when closing the pts the readers will get an EIO error instead of an EOF.  This is what we are getting on Linux during the `ExpectEOF`:
```go
&os.PathError{
    Op:   "read",
    Path: "/dev/ptmx",
    Err:  syscall.Errno(0x5),
}
```

I found this relevant:
https://github.com/kr/pty/issues/21#issuecomment-129381749

If we close the ptm (instead of the pts) then we get this error from ExpectEOF:
```go
&os.PathError{
    Op:   "read",
    Path: "/dev/ptmx",
    Err:  &errors.errorString{s:"file already closed"},
}
```
At this point I am not sure how we can ensure an actual EOF is sent to the goroutine.

File: expect.go [hunk 2]
```diff
@@ -59,11 +59,28 @@ func (c *Console) Expect(opts ...ExpectOpt) (string, error) {
...
 
 	var matcher Matcher
+	var err error
+
+	defer func() {
+		if len(c.opts.ExpectObservers) < 1 {
+			return
+		}
+		b := ""
+		if buf != nil {
```
Comment (hinshun) [inline expect.go]: Is this necessary?

**Pull request merged**